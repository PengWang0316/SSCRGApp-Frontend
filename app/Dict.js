export default {
  en: {
    login: 'login',
    logout: 'logout',
    test: 'test',
    other: 'other',
    appName: 'SSCRG CLUB',
    homePageContent: 'This is the Home Page',
    testPageContent: 'This is the Test Page',
  },
  zh: {
    login: '登录',
    logout: '登出',
    test: '测试',
    other: '其他',
    appName: 'SSCRG 俱乐部',
    homePageContent: '这是首页',
    testPageContent: '这是测试页',
  },
};
