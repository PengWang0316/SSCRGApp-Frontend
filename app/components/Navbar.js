import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import {
  MenuItem, Menu, IconButton, Hidden, Button, Typography, Toolbar, AppBar, Avatar,
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import MenuIcon from '@material-ui/icons/Menu';
import orange from '@material-ui/core/colors/orange';
// import { Auth } from 'aws-amplify';
import I18n from '@kevinwang0316/i18n';

import {
  HOME_PAGE_URL, TEST_PAGE_URL, SIGNIN_PAGE_URL,
} from '../config';
import * as UserActions from '../actions/UserActions';

/* istanbul ignore next */
const styles = theme => ({
  link: {
    color: theme.palette.primary.contrastText,
    textDecoration: 'none',
  },
  menuLink: {
    textDecoration: 'none',
  },
  flex1: {
    flex: 1,
  },
  appbar: {
    maxHeight: 55,
    position: 'sticky',
    top: 0,
  },
  avatar: {
    width: 26,
    height: 26,
    marginRight: 8,
    color: '#fff',
    backgroundColor: orange[800],
  },
});

/** Navbar component */
export class Navbar extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    user: PropTypes.object,
    logout: PropTypes.func.isRequired,
    currentAuthenticatedUser: PropTypes.func.isRequired,
  };

  static defaultProps = { user: null };

  state = { anchorEl: null };

  /**
   * Get the authentication user information.
   */
  componentDidMount() {
    // eslint-disable-next-line react/destructuring-assignment
    this.props.currentAuthenticatedUser();
  }

  /**
   * Changing anchorEl state to an click target element or null.
   * @return {null} No Return.
   */
  handleMenuIconClick = ({ currentTarget }) => this.setState(({ anchorEl }) => ({
    anchorEl: anchorEl ? null : currentTarget,
  }));

  /**
   * Showing the login dialog when the user did not login and logout
   * a user if the user has already login.
   */
  handleLoginButtonClick = () => {
    const { history, user, logout } = this.props;
    if (user) {
      logout();
      history.push('/');
    } else history.push(SIGNIN_PAGE_URL);
  }

  /**
   * The render method to render the jsx.
   * @return {jsx} Return jsx.
   */
  render() {
    const { classes, user } = this.props;
    const { anchorEl } = this.state;
    return (
      <AppBar position="static" className={classes.appbar} data-testid="navbar">
        <Toolbar>
          <Link to={HOME_PAGE_URL} className={`${classes.link} ${classes.flex1}`} data-testid="titleLink">
            <Typography variant="h6" color="inherit">{I18n.get('appName')}</Typography>
          </Link>
          <Hidden only="xs">
            <Link to={TEST_PAGE_URL} className={classes.link}>
              <Button color="inherit" data-testid="testButton">{I18n.get('test')}</Button>
            </Link>
            <Button color="inherit" data-testid="otherButton">{I18n.get('other')}</Button>
            <Button color="inherit" onClick={this.handleLoginButtonClick} data-testid="loginButton">
              {user ? (
                <Fragment>
                  <Avatar className={classes.avatar}><Typography color="inherit">{user.nickname.charAt(0)}</Typography></Avatar>
                  <Typography color="inherit">{I18n.get('logout')}</Typography>
                </Fragment>
              ) : I18n.get('login')}
            </Button>
          </Hidden>
          <Hidden only={['xl', 'lg', 'md', 'sm']}>
            <IconButton
              color="inherit"
              aria-label="Menu"
              onClick={this.handleMenuIconClick}
              aria-owns={anchorEl ? 'simple-menu' : null}
              aria-haspopup="true"
              data-testid="navbarDropMenuButton"
            >
              <MenuIcon />
            </IconButton>
            <Menu
              id="simple-menu"
              anchorEl={anchorEl}
              open={Boolean(anchorEl)}
              onClose={this.handleMenuIconClick}
              data-testid="dropDownMenu"
            >
              <MenuItem>
                <Link to={TEST_PAGE_URL} className={classes.menuLink} data-testid="testLink">
                  <Typography color="textPrimary">{I18n.get('test')}</Typography>
                </Link>
              </MenuItem>
              <MenuItem>
                <Typography color="textPrimary">{I18n.get('other')}</Typography>
              </MenuItem>
              <MenuItem onClick={this.handleLoginButtonClick} data-testid="loginMenu">
                {user ? (
                  <Fragment>
                    <Avatar className={classes.avatar}><Typography color="inherit">{user.nickname.charAt(0)}</Typography></Avatar>
                    <Typography color="textPrimary">{I18n.get('logout')}</Typography>
                  </Fragment>
                ) : I18n.get('login')}
              </MenuItem>
            </Menu>
          </Hidden>
        </Toolbar>
      </AppBar>
    );
  }
}

/* istanbul ignore next */
const mapStateToProps = state => ({ user: state.user });
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  currentAuthenticatedUser: () => dispatch(UserActions.currentAuthenticatedUser()),
  logout: () => dispatch(UserActions.logout()),
});

/* Putting the withRouter to the first position because when test code mocks Link
the withRouter also has to be mocked. But it is hard to really return a react
component to satisfy the whole chain call. */
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(withStyles(styles)(Navbar)));
