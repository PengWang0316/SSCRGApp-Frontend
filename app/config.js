export const BASE_URL = ''; // Production server
/* URLS */
export const HOME_PAGE_URL = '/';
export const TEST_PAGE_URL = '/test';
export const SIGNIN_PAGE_URL = '/signin';

export const cognitoConfig = {
  userPoolId: 'us-west-2_H5b19kIOP',
  region: 'us-west-2',
  userPoolWebClientId: '2nr52a5iheoumqcgr47qnlqojq',
  identityPoolId: 'us-west-2:dc3b0c22-b71b-4196-a0e5-91896228a809',
};

export const amplifyAuthSignOption = {
  signUpConfig: {
    hiddenDefaults: ['phone_number', 'email', 'username'],
    signUpFields: [
      {
        label: 'Nickname', key: 'nickname', required: true, type: 'string', displayOrder: 1,
      },
      {
        label: 'Email', key: 'username', required: true, type: 'string', displayOrder: 2,
      },
    ],
  },
  federated: {
    google_client_id: '953200455973-a81rnojmm1ogtudaqtgsd1ge4bris6gt.apps.googleusercontent.com',
    facebook_app_id: '405214103389849',
  },
};
