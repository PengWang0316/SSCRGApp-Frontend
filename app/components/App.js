
import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import importedComponent from 'react-imported-component';
import CssBaseline from '@material-ui/core/CssBaseline';

// import Navbar from './Navbar';
import { HOME_PAGE_URL } from '../config';
import LoadingAnimation from './SharedComponents/LoadingAnimation';

const theme = createMuiTheme({
  typography: {
    useNextVariants: true,
  },
  palette: {
    primary: {
      light: '#9ecde7',
      main: '#6d9cb5',
      dark: '#3e6e85',
      contrastText: '#fff',
    },
    secondary: {
      light: '#497ca4',
      main: '#105075',
      dark: '#002884',
      contrastText: '#fff',
    },
  },
});

/* istanbul ignore next */
const HomePage = importedComponent(() => import(/* webpackChunkName: "HomePageContainer" *//* webpackPrefetch: true */ './containers/HomePageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });
const TestPage = importedComponent(() => import(/* webpackChunkName: "TestPageContainer" *//* webpackPrefetch: true */ './containers/TestPageContainer').catch(err => console.log(err)), { LoadingComponent: LoadingAnimation });

/**
 * The root component that contains the theme, routers, navbar, and login dialog
 */
export default () => (
  <MuiThemeProvider theme={theme}>
    <Router>
      <div>
        <CssBaseline />
        {/* <Navbar /> */}
        <main>
          <Switch>
            <Route exact path={HOME_PAGE_URL} component={HomePage} />
            <Route exact path="/test" component={TestPage} />
            <Route render={() => <p>Not Fount!</p>} />
          </Switch>
        </main>
      </div>
    </Router>
  </MuiThemeProvider>
);
