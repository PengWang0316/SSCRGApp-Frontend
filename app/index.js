import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { AppContainer } from 'react-hot-loader';
import { PersistGate } from 'redux-persist/integration/react';
import I18n from '@kevinwang0316/i18n';

import App from './components/App';
import configureStore from './store/ConfigureStore';
import './styles/index.global.css';
import registerServiceWorker from './registerServiceWorker';
import { DEFAULT_LANGUAGE } from './config';
import dict from './Dict';

const { store, persistor } = configureStore();

// Fallback to english if dict does not have the user's language
I18n.setDefaultLanguage(DEFAULT_LANGUAGE);
I18n.setDictionary(dict);

// ReactDOM.render(
//   <Provider store={store}>
//     <App />
//   </Provider>,
//   document.getElementById('app')
// );
const render = (Component) => {
  ReactDOM.render(
    <AppContainer>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Component />
        </PersistGate>
      </Provider>
    </AppContainer>,
    document.getElementById('app'),
  );
};
render(App);

// Webpack Hot Module Replacement API
if (module.hot) module.hot.accept('./components/App', () => render(App));

// Registering a service worker in the production enviroment.
registerServiceWorker();
