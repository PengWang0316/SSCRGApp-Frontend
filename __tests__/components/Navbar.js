import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import { Navbar } from '../../app/components/Navbar';
import { SIGNIN_PAGE_URL, HOME_PAGE_URL } from '../../app/config';

jest.mock('@material-ui/core/AppBar', () => 'AppBar');
jest.mock('@material-ui/core/Toolbar', () => 'Toolbar');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/Button', () => 'Button');
jest.mock('@material-ui/core/Hidden', () => 'Hidden');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/core/Menu', () => 'Menu');
jest.mock('@material-ui/core/MenuItem', () => 'MenuItem');
jest.mock('@material-ui/core/Avatar', () => 'Avatar');
jest.mock('@material-ui/icons/Menu', () => 'MenuIcon');
jest.mock('react-router-dom', () => ({ Link: 'Link', withRouter: jest.fn() }));
jest.mock('@kevinwang0316/i18n', () => ({ get: key => key }));
jest.mock('aws-amplify', () => ({
  Auth: {
    currentAuthenticatedUser: jest.fn().mockReturnValue(Promise.resolve({ id: 'userId' })),
  },
}));

describe('Navbar', () => {
  const defaultProps = {
    classes: {
      link: 'link', appbar: 'appbar', menuLink: 'menuLink', flex1: 'flex1', avatar: 'avatar',
    },
    logout: jest.fn(),
    currentAuthenticatedUser: jest.fn(),
    user: null,
    history: { push: jest.fn() },
  };
  const getShallowComponent = (props = defaultProps) => shallow(<Navbar {...props} />);

  beforeEach(() => {
    defaultProps.logout.mockClear();
    defaultProps.history.push.mockClear();
    defaultProps.currentAuthenticatedUser.mockClear();
  });

  test('Initial state and componentDidMount', () => {
    const component = getShallowComponent();
    expect(component.state('anchorEl')).toBe(null);
    expect(defaultProps.currentAuthenticatedUser).toHaveBeenCalledTimes(1);
  });

  test('handleMenuIconClick', () => {
    const component = getShallowComponent();
    component.instance().handleMenuIconClick({ currentTarget: 'currentTarget' });
    expect(component.state('anchorEl')).toEqual('currentTarget');
    component.instance().handleMenuIconClick({ currentTarget: 'currentTarget' });
    expect(component.state('anchorEl')).toBe(null);
  });

  test('handleLoginButtonClick', () => {
    const component = getShallowComponent();
    component.instance().handleLoginButtonClick();

    expect(defaultProps.logout).not.toHaveBeenCalled();
    expect(defaultProps.history.push).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).toHaveBeenLastCalledWith(SIGNIN_PAGE_URL);

    component.setProps({ user: { nickname: 'name' } });
    component.instance().handleLoginButtonClick();
    expect(defaultProps.logout).toHaveBeenCalledTimes(1);
    expect(defaultProps.history.push).toHaveBeenCalledTimes(2);
    expect(defaultProps.history.push).toHaveBeenLastCalledWith(HOME_PAGE_URL);
  });

  test('NavBar snapshot without user', () => expect(renderer.create(<Navbar {...defaultProps} />).toJSON()).toMatchSnapshot());

  test('NavBar snapshot with user', () => expect(renderer.create(<Navbar {...{ ...defaultProps, user: { nickname: 'nickname' } }} />).toJSON()).toMatchSnapshot());
});
