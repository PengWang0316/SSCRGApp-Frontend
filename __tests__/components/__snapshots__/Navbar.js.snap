// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar NavBar snapshot with user 1`] = `
<AppBar
  className="appbar"
  data-testid="navbar"
  position="static"
>
  <Toolbar>
    <Link
      className="link flex1"
      data-testid="titleLink"
      to="/"
    >
      <Typography
        color="inherit"
        variant="h6"
      >
        appName
      </Typography>
    </Link>
    <Hidden
      only="xs"
    >
      <Link
        className="link"
        to="/test"
      >
        <Button
          color="inherit"
          data-testid="testButton"
        >
          message
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="otherButton"
      >
        other
      </Button>
      <Button
        color="inherit"
        data-testid="loginButton"
        onClick={[Function]}
      >
        <Avatar
          className="avatar"
        >
          <Typography
            color="inherit"
          >
            n
          </Typography>
        </Avatar>
        <Typography
          color="inherit"
        >
          logout
        </Typography>
      </Button>
    </Hidden>
    <Hidden
      only={
        Array [
          "xl",
          "lg",
          "md",
          "sm",
        ]
      }
    >
      <IconButton
        aria-haspopup="true"
        aria-label="Menu"
        aria-owns={null}
        color="inherit"
        data-testid="navbarDropMenuButton"
        onClick={[Function]}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        anchorEl={null}
        data-testid="dropDownMenu"
        id="simple-menu"
        onClose={[Function]}
        open={false}
      >
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="testLink"
            to="/test"
          >
            <Typography
              color="textPrimary"
            >
              message
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem>
          <Typography
            color="textPrimary"
          >
            other
          </Typography>
        </MenuItem>
        <MenuItem
          data-testid="loginMenu"
          onClick={[Function]}
        >
          <Avatar
            className="avatar"
          >
            <Typography
              color="inherit"
            >
              n
            </Typography>
          </Avatar>
          <Typography
            color="textPrimary"
          >
            logout
          </Typography>
        </MenuItem>
      </Menu>
    </Hidden>
  </Toolbar>
</AppBar>
`;

exports[`Navbar NavBar snapshot without user 1`] = `
<AppBar
  className="appbar"
  data-testid="navbar"
  position="static"
>
  <Toolbar>
    <Link
      className="link flex1"
      data-testid="titleLink"
      to="/"
    >
      <Typography
        color="inherit"
        variant="h6"
      >
        appName
      </Typography>
    </Link>
    <Hidden
      only="xs"
    >
      <Link
        className="link"
        to="/test"
      >
        <Button
          color="inherit"
          data-testid="testButton"
        >
          message
        </Button>
      </Link>
      <Button
        color="inherit"
        data-testid="otherButton"
      >
        other
      </Button>
      <Button
        color="inherit"
        data-testid="loginButton"
        onClick={[Function]}
      >
        login
      </Button>
    </Hidden>
    <Hidden
      only={
        Array [
          "xl",
          "lg",
          "md",
          "sm",
        ]
      }
    >
      <IconButton
        aria-haspopup="true"
        aria-label="Menu"
        aria-owns={null}
        color="inherit"
        data-testid="navbarDropMenuButton"
        onClick={[Function]}
      >
        <MenuIcon />
      </IconButton>
      <Menu
        anchorEl={null}
        data-testid="dropDownMenu"
        id="simple-menu"
        onClose={[Function]}
        open={false}
      >
        <MenuItem>
          <Link
            className="menuLink"
            data-testid="testLink"
            to="/test"
          >
            <Typography
              color="textPrimary"
            >
              message
            </Typography>
          </Link>
        </MenuItem>
        <MenuItem>
          <Typography
            color="textPrimary"
          >
            other
          </Typography>
        </MenuItem>
        <MenuItem
          data-testid="loginMenu"
          onClick={[Function]}
        >
          login
        </MenuItem>
      </Menu>
    </Hidden>
  </Toolbar>
</AppBar>
`;
